.cirrus.yml

```diff
@@ -19,7 +16,14 @@ test_task:
     - $PYTHON -m pip install pip -U
     - $PYTHON -m pip install poetry -U --pre
     - $PYTHON -m poetry config virtualenvs.in-project true
-  deps_script: $PYTHON -m poetry install
+    - $PYTHON -m poetry config --list
+  # todo: cache handling need to be fixed
+  venv_cache:
+    folder: ".venv"
+    fingerprint_script: echo freebsd-$($PYTHON -c "import sys; print('-'.join(str(v) for v in sys.version_info))")-$(sha512 poetry.lock)
+  deps_script: |
+    $PYTHON -m poetry run pip --version >/dev/null 2>&1 || rm -vrf .venv
```

Comment:
I experimented with doing this exactly, and found that `pip` would run but not the test suite. Here be dragons, if you want to go through with this.

Reply:
I am just using it for now to try figure out whats going on. Did you find much else out?

Reply:
Not really -- `pip --version` passed but the test suite broke pretty spectacularly. On Python 3, Python can't find `pytest` in the venv -- and on Python 2.7, it can but something blows up when it tests symlink support? Maybe your deeper knowledge of Poetry/the test suite will help here.

https://cirrus-ci.com/task/5953586152603648
https://cirrus-ci.com/task/6462722447507456
https://cirrus-ci.com/task/4773872587243520

Reply:
Really appreciate your freedback here @neersighted. I am not quite familiar with FreeBSD. 

From what I can tell, the issue probably has to do with how the virtual environment creation is done on FreeBSD systems either due to the system python package itself or how cirrus-ci sets up the build environment. 

The primary difference, outside of platform specifics, in py3.7 and py2.7 virutal environment creation is the use of the built-in `venv` module and external `virtualenv` package respectively. I am aware that the newer versions of behave differently to the built-in module in that on systems like FreeBSD, the lib64 directory is not a symlink to the lib directory. 

In the case of py3.7, a cache hit as no effect (ie. packages are downloaded again) and in the case of py2.7 a cache hit causes the test suite to fail as mentioned above.

Might consider caching the pip and poetry caches instead of the virtual environment for now so that atleast downloads are improved until the venv can be made portable.

One theory, if it is indeed an issue with symlink, is that the cirrus might be using an archiving option that does not like symlinks.

Reply:
I see -- thanks for the insight. I agree that we could keep the pip package cash -- it should speed things up with no downside. On the other hand, most of the time seems to be in disk I/O as these GCP VMs don't seem to exactly be IOPS monsters... The savings may be as little as 20 seconds so I'm not sure it's worth the extra complexity.

Reply:
Created #2315 to handle caching. Thanks for your help @neersighted.